PyMca5/PyMcaGui/io/TiledCatalogSelector.py

```diff
@@ -0,0 +1,145 @@
+        self.client_connected = self.signals.client_connected
+
+    def on_url_focus_in_event(self, event: QEvent):
+        """Handle the event when the URL widget gains focus."""
+        _logger.debug("TiledCatalogSelector.on_url_focus_in_event()...")
+
+        self._url_buffer = None
+
+    def on_url_text_edited(self, new_text: str):
+        """Handle a notification that the URL is being edited."""
+        _logger.debug("TiledCatalogSelector.on_url_text_edited()...")
+
+        self._url_buffer = new_text
+
+    def on_url_editing_finished(self):
+        """Handle a notification that URL editing is complete."""
+        _logger.debug("TiledCatalogSelector.on_url_editing_finished()...")
+
+        try:
+            for validate in self.validators["url"]:
+                validate(self._url_buffer)
+        except ValueError as exception:
+            _logger.info(str(exception))
+            # TODO: Should we re-raise the exception or emit a signal to inform the viewer?
+            return
+        
+        if getattr(self, "_url_buffer", None) is not None:
+            self.url = self._url_buffer
+            self._url_buffer = None
+
+    def on_connect_clicked(self, checked: bool = False):
+        """Handle a button click to connect to the Tiled client."""
+        _logger.debug("TiledCatalogSelector.on_connect_clicked()...")
+
+        try:
+            new_client = self.client_from_url(self.url)
+        except Exception as exception:
+            _logger.info(str(exception))
+            # TODO: Should we re-raise the exception or emit a signal to inform the viewer?
+            return
+
+        if self.client:
+            # TODO: Clean-up previously connected client?
+            ...
+
+        self.client = new_client
+        self.client_connected.emit(self.client.uri, self.client.context.api_uri)
+
+    @staticmethod
+    def client_from_url(url: str):
+        """Create a Tiled client that is connected to the requested URL."""
+        _logger.debug("TiledCatalogSelector.client_from_url()...")
+
+        return from_uri(url)
+
+
+def urlparse(url: str) -> ParseResult:
+    """Re-raise URL parsing errors with an extra custom message."""
+    try:
+        url_parts = _urlparse(url)
+    except ValueError as exception:
+        raise ValueError(f'{url} is not a valid URL.') from exception
+    
+    if not url_parts.scheme:
+        raise ValueError(
+            f'{url} is not a valid URL. URL must include a scheme.'
+        )
+    
+    if not url_parts.netloc:
+        raise ValueError(
+            f'{url} is not a valid URL. URL must include a network location.'
+        )
+    
+    return url_parts
+
+
+def validate_url_syntax(url: str) -> None:
+    """Verify that input string is parseable as a URL."""
+    urlparse(url)
+
+
+def validate_url_scheme(
+    url: str,
+    valid_schemes: Sequence[str] = ("http", "https"),
+) -> None:
+    """Verify that URL scheme is one of 'valid_schemes'."""
```

Comment:
These validators might ultimately not be that useful.  There mostly here to show a concept that we can use.